File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-16.js
class Solution {
    search(nums, target) {
        let l = 0
        let r = nums.length - 1

        while (l <= r) {
            const mid = Math.floor((l + r) / 2)

            if (nums[mid] === target) return mid

            if (nums[mid] <= nums[r]) {
                if (target > nums[mid] && target <= nums[r]) {
                    l = mid + 1
                } else {
                    r = mid - 1
                }
            } else {
                if (target >= nums[l] && target < nums[mid]) {
                    r = mid - 1
                } else {
                    l = mid + 1
                }
            }
        }

        return -1
    }
}

File: Data Structures & Algorithms/eating-bananas/submission-16.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)

        while l <= r:
            mid = (r - l) // 2 + l
            k = self.hoursToEatBananas(piles, mid)

            if k > h:
                l = mid + 1
            else:
                r = mid - 1

        return l
    
    def hoursToEatBananas(self, piles, rate):
        hours = 0
        for p in piles:
            hours += math.ceil(p / rate)

        return hours